### PR TITLE
[kueuectl] Stop/resume cluster queue

### LINKED ISSUE #2136: [kueuectl] Add stop ClusterQueue command
**What would you like to be added**:
Add stop ClusterQueue command on kueuectl.

**Why is this needed**:
To stops admission and execution inside the specified ClusterQueue, possibly limiting the action only to the selected ResourceFlavor.

Design details https://github.com/kubernetes-sigs/kueue/tree/main/keps/2076-kueuectl#stop-clusterqueue.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by vladikkuzn**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds `kueuectl stop clusterqueue`, `kueuectl resume clusterqueue` commands
#### Which issue(s) this PR fixes:

Fixes #2136 and #2137

#### Does this PR introduce a user-facing change?

```release-note
Stop/Resume ClusterQueue (`kueuectl stop clusterqueue`, `kueuectl resume clusterqueue`) commands are available
```

### CODE REVIEW

File: cmd/kueuectl/app/stop/stop_clusterqueue.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *ClusterQueueOptions) stopClusterQueue(cq *v1beta1.ClusterQueue) {
+	if o.KeepAlreadyRunning {
+		cq.Spec.StopPolicy = ptr.To(v1beta1.Hold)
```
Comment (mwielgus) [inline cmd/kueuectl/app/stop/stop_clusterqueue.go:155]: What happens to workloads that are waiting for admission check but are not yet fully started? Are they considered running or not running?

cc: @trasc

File: cmd/kueuectl/app/resume/resume_clusterqueue.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// Run executes the command
+func (o *ClusterQueueOptions) Run(ctx context.Context) error {
+	cq, err := o.Client.ClusterQueues().Get(ctx, o.ClusterQueueName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if cq == nil {
```
Comment (mbobrovskyi) [inline cmd/kueuectl/app/resume/resume_clusterqueue.go:112]: Is it possible this case?

File: test/integration/kueuectl/resume_test.go
Comment (IrvingMg) [inline test/integration/kueuectl/resume_test.go:1]: Is it possible to use `DescribeTable`?  If so, maybe we can do the same for stop tests

File: cmd/kueuectl/app/resume/resume_clusterqueue.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+type ClusterQueueOptions struct {
+	PrintFlags *genericclioptions.PrintFlags
+
+	ClusterQueueName string
+
+	Client kueuev1beta1.KueueV1beta1Interface
+
+	PrintObj printers.ResourcePrinterFunc
+
```
Comment (trasc) [inline cmd/kueuectl/app/resume/resume_clusterqueue.go:51]: nit: why so many empty lines?

File: cmd/kueuectl/app/resume/resume_clusterqueue.go (hunk 2 above)
Comment (trasc) [reply cmd/kueuectl/app/resume/resume_clusterqueue.go:112]: likely no.

File: cmd/kueuectl/app/resume/resume_clusterqueue.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return err
+	}
+
+	return o.PrintObj(cq, o.Out)
+}
+
+func (o *ClusterQueueOptions) resumeClusterQueue(cq *v1beta1.ClusterQueue) {
+	cq.Spec.StopPolicy = ptr.To(v1beta1.None)
+}
```
Comment (trasc) [inline cmd/kueuectl/app/resume/resume_clusterqueue.go:136]: We should avoid one statement functions.

File: cmd/kueuectl/app/stop/helpers.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+package stop
+
+import (
+	"github.com/spf13/cobra"
+)
+
+func addKeepAlreadyRunningFlagVar(cmd *cobra.Command, p *bool) {
+	cmd.Flags().BoolVar(p, "keep-already-running", false,
+		"Indicates whether to keep already running local queues.")
```
Comment (trasc) [inline cmd/kueuectl/app/stop/helpers.go:25]:
```suggestion
		"Indicates whether to keep already running workloads.")
```

File: cmd/kueuectl/app/stop/stop_clusterqueue.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	PrintFlags *genericclioptions.PrintFlags
+
+	ClusterQueueName   string
+	KeepAlreadyRunning bool
+
+	Client kueuev1beta1.KueueV1beta1Interface
+
+	PrintObj printers.ResourcePrinterFunc
+
+	genericiooptions.IOStreams
```
Comment (trasc) [inline cmd/kueuectl/app/stop/stop_clusterqueue.go:53]: nit: empty lines

File: cmd/kueuectl/app/stop/stop_clusterqueue.go (hunk 1 above)
Comment (trasc) [reply cmd/kueuectl/app/stop/stop_clusterqueue.go:139]: Yes, the workloads that are only reserving louse the reservation immediately.

File: test/integration/kueuectl/resume_test.go [hunk 7]
```diff
@@ -78,4 +79,72 @@ var _ = ginkgo.Describe("Kueuectl Resume", ginkgo.Ordered, ginkgo.ContinueOnFail
 			})
 		})
 	})
+
+	ginkgo.When("Resuming a ClusterQueue from HoldAndDrain StopPolicy", func() {
+		ginkgo.It("Should resume a ClusterQueue and drain workloads", func() {
```
Comment (trasc) [inline test/integration/kueuectl/resume_test.go:84]:
```suggestion
		ginkgo.It("Should resume a ClusterQueue", func() {
```

File: test/integration/kueuectl/resume_test.go [hunk 8]
```diff
@@ -78,4 +79,72 @@ var _ = ginkgo.Describe("Kueuectl Resume", ginkgo.Ordered, ginkgo.ContinueOnFail
...
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), cq)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.None)))
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+		})
+	})
+
+	ginkgo.When("Resuming a ClusterQueue from Hold StopPolicy", func() {
+		ginkgo.It("Should resume a ClusterQueue and drain workloads", func() {
```
Comment (trasc) [inline test/integration/kueuectl/resume_test.go:118]:
```suggestion
		ginkgo.It("Should resume a ClusterQueue", func() {
```

File: test/integration/kueuectl/resume_test.go [hunk 9]
```diff
@@ -78,4 +79,72 @@ var _ = ginkgo.Describe("Kueuectl Resume", ginkgo.Ordered, ginkgo.ContinueOnFail
...
+
+				kueuectl.SetArgs([]string{"resume", "clusterqueue", cq.Name})
+				err := kueuectl.Execute()
+				gomega.Expect(err).NotTo(gomega.HaveOccurred(), "%s: %s", err, output)
+			})
+
+			ginkgo.By("Check that the ClusterQueue is successfully resumed", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), cq)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.None)))
```
Comment (trasc) [inline test/integration/kueuectl/resume_test.go:111]:
```suggestion
					g.Expect(ptr.Deref(cq.Spec.StopPolicy, v1beta1.None)).Should(gomega.Equal(v1beta1.None))
```

File: test/integration/kueuectl/resume_test.go [hunk 10]
```diff
@@ -78,4 +79,72 @@ var _ = ginkgo.Describe("Kueuectl Resume", ginkgo.Ordered, ginkgo.ContinueOnFail
...
```
Comment (trasc) [inline test/integration/kueuectl/resume_test.go:145]:
```suggestion
					g.Expect(ptr.Deref(cq.Spec.StopPolicy, v1beta1.None)).Should(gomega.Equal(v1beta1.None))
```

File: test/integration/kueuectl/resume_test.go [hunk 11]
```diff
@@ -78,4 +79,72 @@ var _ = ginkgo.Describe("Kueuectl Resume", ginkgo.Ordered, ginkgo.ContinueOnFail
...
+			cq := testing.MakeClusterQueue("cq-2").StopPolicy(v1beta1.Hold).Obj()
+			ginkgo.By("Create a ClusterQueue", func() {
+				gomega.Expect(k8sClient.Create(ctx, cq)).To(gomega.Succeed())
+			})
+
+			createdClusterQueue := &v1beta1.ClusterQueue{}
+			ginkgo.By("Get created ClusterQueue", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), createdClusterQueue)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.Hold)))
```
Comment (trasc) [inline test/integration/kueuectl/resume_test.go:128]:
```suggestion
					g.Expect(ptr.Deref(cq.Spec.StopPolicy, v1beta1.None)).Should(gomega.Equal(v1beta1.Hold))
```

File: test/integration/kueuectl/resume_test.go [hunk 12]
```diff
@@ -78,4 +79,72 @@ var _ = ginkgo.Describe("Kueuectl Resume", ginkgo.Ordered, ginkgo.ContinueOnFail
...
+			cq := testing.MakeClusterQueue("cq-1").StopPolicy(v1beta1.HoldAndDrain).Obj()
+			ginkgo.By("Create a ClusterQueue", func() {
+				gomega.Expect(k8sClient.Create(ctx, cq)).To(gomega.Succeed())
+			})
+
+			createdClusterQueue := &v1beta1.ClusterQueue{}
+			ginkgo.By("Get created ClusterQueue", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), createdClusterQueue)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.HoldAndDrain)))
```
Comment (trasc) [inline test/integration/kueuectl/resume_test.go:94]:
```suggestion
					g.Expect(ptr.Deref(cq.Spec.StopPolicy, v1beta1.None)).Should(gomega.Equal(v1beta1.HoldAndDrain))
```

File: test/integration/kueuectl/stop_test.go [hunk 13]
```diff
@@ -78,4 +79,70 @@ var _ = ginkgo.Describe("Kueuectl Stop", ginkgo.Ordered, ginkgo.ContinueOnFailur
...
+
+				kueuectl.SetArgs([]string{"stop", "clusterqueue", cq.Name})
+				err := kueuectl.Execute()
+				gomega.Expect(err).NotTo(gomega.HaveOccurred(), "%s: %s", err, output)
+			})
+
+			ginkgo.By("Check that the ClusterQueue is successfully stopped", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), cq)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.HoldAndDrain)))
```
Comment (trasc) [inline test/integration/kueuectl/stop_test.go:111]: deref and check against value.

File: test/integration/kueuectl/stop_test.go [hunk 14]
```diff
@@ -78,4 +79,70 @@ var _ = ginkgo.Describe("Kueuectl Stop", ginkgo.Ordered, ginkgo.ContinueOnFailur
...
+
+				kueuectl.SetArgs([]string{"stop", "clusterqueue", cq.Name, "--keep-already-running"})
+				err := kueuectl.Execute()
+				gomega.Expect(err).NotTo(gomega.HaveOccurred(), "%s: %s", err, output)
+			})
+
+			ginkgo.By("Check that the ClusterQueue is successfully stopped", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), cq)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.Hold)))
```
Comment (trasc) [inline test/integration/kueuectl/stop_test.go:143]: deref and check against value.

File: test/integration/kueuectl/stop_test.go [hunk 15]
```diff
@@ -78,4 +79,70 @@ var _ = ginkgo.Describe("Kueuectl Stop", ginkgo.Ordered, ginkgo.ContinueOnFailur
...
+
+			ginkgo.By("Check that the ClusterQueue is successfully stopped", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(cq), cq)).To(gomega.Succeed())
+					g.Expect(cq.Spec.StopPolicy).Should(gomega.Equal(ptr.To(v1beta1.HoldAndDrain)))
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+		})
+
+		ginkgo.It("Should stop a ClusterQueue and finish workloads", func() {
```
Comment (trasc) [inline test/integration/kueuectl/stop_test.go:116]:
```suggestion
		ginkgo.It("Should stop a ClusterQueue and let the admitted workloads finish", func() {
```
or something

**Review by tenzen-y — COMMENTED** [review]
Thanks LGTM
/approve

**Comment by tenzen-y** [comment]
leaving lgtm on @trasc

**Review by trasc — COMMENTED** [review]
/lgtm

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
CLI: Add commands to stop and Resume a ClusterQueue
```

**Pull request merged**